internal/gcsutil/util.go

```diff
@@ -0,0 +1,124 @@
+// Copyright 2022 The Project Oak Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package gcsutil contains utility functions for Google Cloud Storage data access.
+package gcsutil
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"strings"
+
+	"cloud.google.com/go/storage"
+	"google.golang.org/api/iterator"
+)
+
+// GCSClient contains a Google Cloud Storage client.
+type GCSClient struct {
+	Client *storage.Client
+}
+
+// NewClient creates and returns a new GCSClient.
+func NewClient(ctx context.Context) (*GCSClient, error) {
+	client, err := storage.NewClient(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("could not create a new Google Cloud Storage client: %v", err)
+	}
+	defer client.Close()
+	gcsClient := GCSClient{
+		Client: client,
+	}
+	return &gcsClient, nil
+}
+
+// ListBlobs returns all the objects paths in a Google Cloud Storage bucket
+// under a given relative path.
+func (client *GCSClient) ListBlobs(ctx context.Context, bucketName string, relativePath string) ([]string, error) {
+	query := &storage.Query{Prefix: relativePath}
+	objects := client.Client.Bucket(bucketName).Objects(ctx, query)
+	var blobs []string
+	for {
+		attrs, err := objects.Next()
+		if err == iterator.Done {
+			break
+		}
+		if err != nil {
+			return nil, fmt.Errorf("could not fetch object from bucket: %v", err)
+		}
+		blobs = append(blobs, attrs.Name)
+	}
+	return blobs, nil
+}
+
+// GetLogs returns all the log-files paths in a Google Cloud Storage bucket
+// under a given relative path.
+func (client *GCSClient) GetLogs(ctx context.Context, bucketName string, relativePath string) ([]string, error) {
+	blobs, err := client.ListBlobs(ctx, bucketName, relativePath)
```

Comment:
If I understand correctly, we first list all the blobs and then we iterate over all the blobs again to process them. 

Does it make sense to do this in two steps? Can't we do it in one? 

Especially as we don't seem to need all the blobs anyway.

// EDIT: I just saw that we need ListBlobs somewhere else too. So you can ignore my comment or think about it :)

Reply:
Thanks, I agree. I updated it to avoid having two steps.